drivers/network/imx/ethernet.c

```diff
@@ -102,272 +79,150 @@ static void update_ring_slot(
 
-    /* Only get notified if more rx bufs are available if hw ring not full */
-    if (!hw_ring_full(ring)) {
-        rx_ring.free_ring->notify_reader = true;
-    } else {
-        rx_ring.free_ring->notify_reader = false;
-    }
-
-    THREAD_MEMORY_FENCE();
-
-    if (!ring_empty(rx_ring.free_ring) && !hw_ring_full(ring)) {
-        rx_ring.free_ring->notify_reader = false;
-        goto fill_rx_bufs_;
+        if (!ring_empty(rx_ring.free_ring) && !hw_ring_full(&rx, RX_COUNT)) {
+            cancel_signal(rx_ring.free_ring);
+            reprocess = true;
+        }
     }
 
-    if (!(hw_ring_empty(ring))) {
-        /* Make sure rx is enabled*/
+    if (!(hw_ring_empty(&rx, RX_COUNT))) {
+        /* Ensure rx IRQs are enabled */
         eth->rdar = RDAR_RDAR;
-        if (!(irq_mask & NETIRQ_RXF))
-            enable_irqs(eth, IRQ_MASK);
+        if (!(irq_mask & NETIRQ_RXF)) enable_irqs(IRQ_MASK);
     } else {
-        enable_irqs(eth, NETIRQ_TXF | NETIRQ_EBERR);
+        enable_irqs(NETIRQ_TXF | NETIRQ_EBERR);
     }
 }
 
-static void
-handle_rx(volatile struct enet_regs *eth)
+static void rx_return(void)
 {
-    ring_ctx_t *ring = &rx;
     bool packets_transferred = false;
-    unsigned int read = ring->read;
-
-    if (ring_full(rx_ring.used_ring))
-    {
-        /*
-         * we can't process any packets because the queues are full
-         * so disable Rx irqs.
-         */
-        enable_irqs(eth, NETIRQ_TXF | NETIRQ_EBERR);
-        __sync_synchronize();
-        rx_ring.used_ring->notify_writer = true;
-        return;
-    }
-
-    /*
-     * Dequeue until either:
-     *   we run out of filled buffers in the NIC ring
-     *   we run out of slots in the upstream Rx ring.
-     */
-    while (!hw_ring_empty(ring) && !ring_full(rx_ring.used_ring)) {
-        volatile struct descriptor *d = &(ring->descr[read]);
-        /*
-         * If a slot is still marked as empty we are done.
-         * Still need to signal upstream if
-         * we've dequeued anything.
-         */
-        if (d->stat & RXD_EMPTY) {
-            break;
-        }
-
-        void *cookie = ring->cookies[read];
-        /* Go to next buffer, handle roll-over. */
-        if (++read == ring->cnt) {
-            read = 0;
-        }
-        THREAD_MEMORY_RELEASE();
-        ring->read = read;
-
-        buff_desc_t *desc = (buff_desc_t *)cookie;
-        int err = enqueue_used(&rx_ring, desc->encoded_addr, d->len, desc->cookie);
-        if (err) {
-            print("ETH|ERROR: Failed to enqueue to RX used ring\n");
-        }
+    while (!hw_ring_empty(&rx, RX_COUNT)) {
+        /* If buffer slot is still empty, we have processed all packets the device has filled */
+        volatile struct descriptor *d = &(rx.descr[rx.head]);
+        if (d->stat & RXD_EMPTY) break;
+
+        buff_desc_t buffer = rx.descr_mdata[rx.head];
+        buffer.len = d->len;
+        int err __attribute__((unused)) = enqueue_used(&rx_ring, buffer);
```

Comment:
Can you remove all instances of `__attribute__((unused))`, it should be necessary anymore since I changed assert in `util.h` to include it.

Reply:
Done